src/ImpactX.cpp

```diff
@@ -65,8 +65,26 @@ namespace impactx
     void ImpactX::MakeNewLevelFromScratch (int lev, amrex::Real time, const amrex::BoxArray& ba,
                                           const amrex::DistributionMapping& dm)
     {
-        // todo data_mf.define(ba, dm, 1, 0);
-        amrex::ignore_unused(lev, time, ba, dm);
+        amrex::ignore_unused(time, ba, dm);
+
+        // set human-readable tag for each MultiFab
+        auto const tag = [lev]( std::string tagname ) {
+            tagname.append("[l=").append(std::to_string(lev)).append("]");
+            return amrex::MFInfo().SetTag(std::move(tagname));
+        };
+
+        // charge (rho) mesh
+        amrex::BoxArray cba = ba;
+        // for MR levels (TODO):
+        //cba.coarsen(refRatio(lev - 1));
+
+        auto const rho_nodal_flag = amrex::IntVect::TheNodeVector();
+        int const num_components_rho = 1;
+        // as in WarpX: this is likely too large and something like shape/2+1 would do.
+        int const num_guards_rho = m_particle_container->GetParticleShape() + 1;
```

Comment:
Oh, forgot this one.
```suggestion
        // as in WarpX: this is likely too large and something like shape/2+1 would do.
        int const num_guards_rho = m_particle_container->GetParticleShape()/2 + 1;
```

Reply:
I tried `shape/2 + 1` and even `10`.
Whatever I do, the lower bound of `algo.particle_shape = 3` seems to get out-of-bounds at the moment.

Reply:
fixed xyzmin